### PR TITLE
shirley / s1340 add actions

### PR DESCRIPTION
**Pull request by shirleygordon-blink**

### CODE REVIEW

File: custom_actions/actions.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+	applicationName := params["Application Name"]
+	applicationID, err := getApplicationIDByName(ctx, requestUrl, applicationName, request.Timeout)
+	if err != nil {
+		return &plugin.ExecuteActionResponse{ErrorCode: consts.Error, Result: []byte(err.Error())}, nil
+	}
+	query, limit, after := params["Query"], params["Limit"], params["After"]
+	url := requestUrl+"/api/v1/apps/" + applicationID + "/users?query_scope=" + url.QueryEscape(query) + "&limit=" + url.QueryEscape(limit) + "&after=" + url.QueryEscape(after)
+	req, _ := http.NewRequest(http.MethodGet, url, nil)
+	return execRequest(ctx, req, request.Timeout)
+}
```
Comment (itsofirblink) [inline custom_actions/actions.go:57]: new line :)

File: custom_actions/actions.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+	res.Result = response.Body
+	if response.StatusCode >= 200 && response.StatusCode < 300 {
+		return res, nil
+	}
+	res.ErrorCode = consts.Error
+	return res, nil
+}
+
+func listApplicationUsersByName(ctx *plugin.ActionContext, request *plugin.ExecuteActionRequest) (*plugin.ExecuteActionResponse, error) {
+	requestUrl, _ := openapi_sdk.GetRequestUrl(ctx, PluginName)
```
Comment (itsofirblink) [inline custom_actions/actions.go:43]: handle error :)

File: cmd/main/main.go [hunk 3]
```diff
@@ -64,11 +59,11 @@ func main() {
 
 // ValidateCredentials test if the provided credentials are correct.
 func ValidateCredentials(ctx *plugin.ActionContext) (bool, []byte) {
-	requestUrl, err := openapi_sdk.GetRequestUrl(ctx, PluginName)
+	requestUrl, err := openapi_sdk.GetRequestUrl(ctx, actions.PluginName)
 
 	req, _ := http.NewRequest(http.MethodGet, requestUrl+"/api/v1/apps", nil)
```
Comment (itsofirblink) [inline cmd/main/main.go:64]: handle error

File: custom_actions/helpers.go [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+	"fmt"
+	openapi "github.com/blinkops/blink-openapi-sdk/plugin"
+	"github.com/blinkops/blink-sdk/plugin"
+	log "github.com/sirupsen/logrus"
+	"net/http"
+	"net/url"
+)
+
+func getApplicationIDByName(ctx *plugin.ActionContext, requestUrl string, applicationName string, timeout int32) (string, error) {
+	req, _ := http.NewRequest(http.MethodGet, requestUrl+"/api/v1/apps?q=" + url.QueryEscape(applicationName), nil)
```
Comment (itsofirblink) [inline custom_actions/helpers.go:14]: handle error :)

File: custom_actions/actions.go [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+	params, err := request.GetParameters()
+	if err != nil {
+		return nil, err
+	}
+	applicationName := params["Application Name"]
```
Comment (itsofirblink) [inline custom_actions/actions.go:48]: handle if its not provided 
```golang
applicationName, ok := ...
if !ok {
return nil, errors.New(...)
}
```

File: custom_actions/actions.go [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+	applicationID, err := getApplicationIDByName(ctx, requestUrl, applicationName, request.Timeout)
+	if err != nil {
+		return &plugin.ExecuteActionResponse{ErrorCode: consts.Error, Result: []byte(err.Error())}, nil
+	}
+	query, limit, after := params["Query"], params["Limit"], params["After"]
```
Comment (itsofirblink) [inline custom_actions/actions.go:53]: same here?

File: custom_actions/actions.go [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+	url := requestUrl+"/api/v1/apps/" + applicationID + "/users?query_scope=" + url.QueryEscape(query) + "&limit=" + url.QueryEscape(limit) + "&after=" + url.QueryEscape(after)
```
Comment (itsofirblink) [inline custom_actions/actions.go:54]: if there is no `query` / `limit` / `after` are you sure its going to work?

File: custom_actions/actions/ListApplicastionUsersByName.yaml [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+    index: 2
+  "Limit":
+    default: -1
+    description: "Specifies the number of results for a page (maximum 500)"
+    type: "int32"
+    index: 3
+  "After":
+    description: "Specifies the pagination cursor for the next page of assignments"
+    type: "string"
+    index: 4
```
Comment (itsofirblink) [inline custom_actions/actions/ListApplicastionUsersByName.yaml:22]: new line

File: custom_actions/helpers.go [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+	response, err := openapi.ExecuteRequest(ctx, req, PluginName, Prefixes, HeaderAlias, timeout, nil)
+	if err != nil {
+		log.Error(err.Error())
+		return "", err
+	}
+	var apps []struct {
+		ApplicationID string `json:"id"`
+	}
+	err = json.Unmarshal(response.Body, &apps)
+	if err != nil {
```
Comment (itsofirblink) [inline custom_actions/helpers.go:24]: in-line :)

File: custom_actions/helpers.go [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+		return "", err
+	}
+	if len(apps) == 0 {
+		return "", fmt.Errorf("the application name input did not match any application")
+	}
+	if len(apps) > 1 {
+		return "", fmt.Errorf("the application name input matched more than one application")
+	}
+	return apps[0].ApplicationID, nil
+}
```
Comment (itsofirblink) [inline custom_actions/helpers.go:35]: new line

**Review by itsofirblink — CHANGES_REQUESTED** [review]

File: custom_actions/actions.go (hunk 7 above)
Comment (shirleygordon-blink) [reply custom_actions/actions.go:54]: Yup, tested it

File: custom_actions/actions.go (hunk 5 above)
Comment (shirleygordon-blink) [reply custom_actions/actions.go:48]: the key will always appear in the map even if the user didn't provide any input, it'll just be an empty string - in this case I'll just check if the application name is empty

File: custom_actions/actions.go (hunk 6 above)
Comment (shirleygordon-blink) [reply custom_actions/actions.go:53]: the key will always appear in the map even if the user didn't provide any input, it'll just be an empty string, in this case it doesn't matter if it's empty, it'll still work

**Review by itsofirblink — APPROVED** [review]

**Pull request merged**